contracts/src/FunctionGateway.sol

```diff
@@ -2,246 +2,270 @@
-        } else if (_outputsRoot != keccak256(abi.encode(_outputHashes))) {
-            revert OutputsRootMismatch(_outputsRoot, _outputHashes);
-        } else if (_verificationKeyRoot != keccak256(abi.encode(verificationKeyHashes))) {
-            revert VerificationKeysRootMismatch(_verificationKeyRoot, verificationKeyHashes);
+        // Assert that the callback hash is unfilfilled.
+        if (requests[_nonce] != requestHash) {
+            revert InvalidRequest(_nonce, requests[_nonce], requestHash);
         }
 
-        // Update the requests.
-        for (uint256 i = 0; i < _requestIds.length; i++) {
-            bytes32 requestId = _requestIds[i];
-            requests[requestId].proofFulfilled = true;
-            requests[requestId].outputHash = _outputHashes[i];
-        }
+        // Compute the output hash.
+        bytes32 outputHash = sha256(_output);
 
-        // Verify the aggregate proof.
-        address aggregationVerifier = verifiers[AGGREGATION_FUNCTION_ID];
-        if (!IFunctionVerifier(aggregationVerifier).verify(_inputsRoot, _outputsRoot, _aggregateProof)) {
-            revert InvalidProof(address(aggregationVerifier), _inputsRoot, _outputsRoot, _aggregateProof);
-        }
+        // Verify the proof.
+        _verify(_functionId, _inputHash, outputHash, _proof);
 
-        emit ProofBatchFulfilled(
-            _requestIds, _aggregateProof, _inputsRoot, _outputHashes, _outputsRoot, _verificationKeyRoot
+        // Execute the callback.
+        isCallback = true;
+        (bool status, ) = _callbackAddress.call(
+            abi.encodeWithSelector(_callbackSelector, _output, _context)
         );
-    }
+        isCallback = false;
 
-    /// @dev Fulfills the callback for a request.
-    /// @param _requestId The id of the request to be fulfilled.
-    /// @param _output The output of the proof.
-    /// @param _context The context of the runtime.
-    function callback(bytes32 _requestId, bytes memory _output, bytes memory _context) external {
-        // Do some sanity checks.
-        FunctionRequest storage r = requests[_requestId];
-        if (r.callbackFulfilled) {
-            revert CallbackAlreadyFulfilled(_requestId);
-        } else if (r.callbackAddress == address(0)) {
-            revert RequestNotFound(_requestId);
-        } else if (r.contextHash != keccak256(_context)) {
-            revert ContextMismatch(_requestId, _context);
-        } else if (r.outputHash != sha256(_output)) {
-            revert OutputMismatch(_requestId, _output);
-        } else if (!r.proofFulfilled) {
-            revert ProofNotFulfilled(_requestId);
+        // If the callback failed, revert.
+        if (!status) {
+            revert CallbackFailed(_callbackSelector, _output, _context);
         }
 
-        // Update the request.
-        r.callbackFulfilled = true;
+        // Delete the callback hash for a gas refund.
+        delete requests[_nonce];
```

Comment:
I think here you have to delete the requests[_nonce] before the external call for re-entrancy reasons?